docs/experiments-monot5-tpu.md

```diff
@@ -0,0 +1,214 @@
+                                      --t5_input_ids ${DATA_DIR}/query_doc_pair_ids.dev.small.tsv
+```
+We will get two output files here:
+- `query_doc_pairs.dev.small.txt`: The query-doc pairs for monoT5 input.
+- `query_doc_pair_ids.dev.small.tsv`: The `query_id`s and `doc_id`s that mapping to the query-doc pairs. We will use this to map monoT5 output scores back to query-doc pairs.
+
+Note that there will be memory issue if the monoT5 input file is larger than 1,000,000 lines. Thus, we will split the input file into multiple files.
+
+```
+split --suffix-length 3 --numeric-suffixes --lines 999900 ${DATA_DIR}/query_doc_pairs.dev.small.txt ${DATA_DIR}/query_doc_pairs.dev.small.txt
+```
+
+For `query_doc_pairs.dev.small.txt`, we will get 7 files after split. i.e. (`query_doc_pairs.dev.small.txt000` to `query_doc_pairs.dev.small.txt006`)
+
+Then copy these input files to Google Storage. TPU inference will read data directly from `gs`
+```
+export GS_FOLDER=<google storage folder to store input/output data>
+gsutil cp ${DATA_DIR}/query_doc_pairs.dev.small.txt??? ${GS_FOLDER}
+```
+
+## Start a VM with TPU on Google Cloud
+
+Define environment variables.
+```
+export PROJECT_NAME=<gcloud project name>
+export PROJECT_ID=<gcloud project id>
+export INSTANCE_NAME=<name of vm to create>
+export TPU_NAME=<name of tpu to create>
+```
+
+Create the VM.
+```
+gcloud beta compute --project=${PROJECT_NAME} instances create ${INSTANCE_NAME} --zone=europe-west4-a --machine-type=n1-standard-4 --subnet=default --network-tier=PREMIUM --maintenance-policy=MIGRATE --service-account=${PROJECT_ID}-compute@developer.gserviceaccount.com  --scopes=https://www.googleapis.com/auth/cloud-platform --image=debian-9-stretch-v20191014 --image-project=debian-cloud --boot-disk-size=200GB --boot-disk-type=pd-standard --boot-disk-device-name=${INSTANCE_NAME} --reservation-affinity=any
+```
+
+After the VM created, we can `ssh` to the machine.  
+Then create a TPU.
+
+```
+curl -O https://dl.google.com/cloud_tpu/ctpu/latest/linux/ctpu && chmod a+x ctpu
+
+./ctpu up --name=${TPU_NAME} --project=${PROJECT_NAME} --zone=europe-west4-a --tpu-size=v3-8 --tpu-only --tf-version=1.15 --noconf
+```
+
+## Setup environment on VM
+Install required tools.
+```
+sudo apt-get update
+sudo apt-get install git gcc screen --yes
+```
+
+Install [Miniconda](https://docs.conda.io/en/latest/miniconda.html).
+```
+curl -O https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
+bash ./Miniconda3-latest-Linux-x86_64.sh
+```
+When installation finished, do:
+```
+source ~/.bashrc
+```
+
+Then create a Python virtual environment for the experiments. 
+```
+conda init
+conda create --y --name py36 python=3.6
+conda activate py36
+```
+Install dependencies.
+```
+conda install -c conda-forge httptools jsonnet --yes
+pip install tensorflow==1.15 tensorflow-text==1.15 t5[gcp]
+git clone https://github.com/ronakice/mesh.git
```

Comment:
Can we use the one in `https://github.com/castorini/mesh`? Also, make sure it is working? Its been a long time since I used it..

Reply:
@kelvin-jiang made some change for the mesh to fit tf2.0 recently. I think it is safer for him to update this part later?

Reply:
I made the same changes (that were in ronak's/castorini's fork) on top of the latest https://github.com/tensorflow/mesh.git, which allows us to use TF 2.0 like so:

```
conda init
conda create --y --name py36 python=3.6
conda activate py36
conda install -c conda-forge httptools jsonnet --yes
pip install tensorflow tensorflow-text t5[gcp]
git clone -b tf2 https://github.com/kelvin-jiang/mesh.git
pip install --editable mesh
./ctpu up --name=tpu-$URA-$EXPNO --project=$PROJECT --zone=europe-west4-a --tpu-size=v3-8 --tpu-only --noconf
```

I'm not sure if we've all adopted/used this, so I don't know if we should include it in these docs?

Reply:
This is great! Could you please merge your code into https://github.com/castorini/mesh?

Reply:
Don't have write access to repo, I opened a PR instead! Let me know when it's merged (or if you have any concerns) and I can try running with it once.

Reply:
Awesome, thanks! @MXueguang, could you please update the instructions to point to the castorini repo?

Reply:
@rodrigonogueira4 do we still keep tf1.15 for now or change to tf2.0

Reply:
@kelvin-jiang , I am getting this error by using the new mesh, Have you seen this error before?
```
    return fn(*new_args, **new_kwargs)
  File "/home/mxg123mxg456mxg/mesh/mesh_tensorflow/transformer/utils.py", line 1205, in decode_from_file
    estimator, input_fn, vocabulary, checkpoint_path=checkpoint_path)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/gin/config.py", line 1078, in gin_wrapper
    utils.augment_exception_message_and_reraise(e, err_str)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/gin/utils.py", line 49, in augment_exception_message_and_reraise
    six.raise_from(proxy.with_traceback(exception.__traceback__), None)
  File "<string>", line 3, in raise_from
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/gin/config.py", line 1055, in gin_wrapper
    return fn(*new_args, **new_kwargs)
  File "/home/mxg123mxg456mxg/mesh/mesh_tensorflow/transformer/utils.py", line 1088, in decode
    for i, result in enumerate(result_iter):
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3132, in predict
    rendezvous.raise_errors()
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/error_handling.py", line 150, in raise_errors
    six.reraise(typ, value, traceback)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/six.py", line 703, in reraise
    raise value
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3126, in predict
    yield_single_examples=yield_single_examples):
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 613, in predict
    self.config)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 2921, in _call_model_fn
    config)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 1163, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3201, in _model_fn
    input_holders.generate_infeed_enqueue_ops_and_dequeue_fn())
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 1465, in generate_infeed_enqueue_ops_and_dequeue_fn
    self._invoke_input_fn_and_record_structure())
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 1520, in _invoke_input_fn_and_record_structure
    num_hosts))
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 1126, in generate_broadcast_enqueue_ops_fn
    inputs = _Inputs.from_input_fn(input_fn(user_context))
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3055, in _input_fn
    return input_fn(**kwargs)
  File "/home/mxg123mxg456mxg/mesh/mesh_tensorflow/transformer/utils.py", line 1196, in input_fn
    dataset = tf.data.Dataset.from_tensor_slices({"inputs": all_input_ids})
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 2421, in from_tensor_slices
    return DatasetV1Adapter(DatasetV2.from_tensor_slices(tensors))
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 682, in from_tensor_slices
    return TensorSliceDataset(tensors)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 3001, in __init__
    element = structure.normalize_element(element)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/data/util/structure.py", line 115, in normalize_element
    ops.convert_to_tensor(t, name="component_%d" % i))
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1499, in convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/framework/tensor_conversion_registry.py", line 52, in _default_conversion_function
    return constant_op.constant(value, dtype, name=name)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/framework/constant_op.py", line 264, in constant
    allow_broadcast=True)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/framework/constant_op.py", line 286, in _constant_impl
    "Const", [], [dtype_value.type], attrs=attrs, name=name).outputs[0]
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3485, in _create_op_internal
    op_def=op_def)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1975, in __init__
    control_input_ops, op_def)
  File "/home/mxg123mxg456mxg/miniconda3/envs/py36/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1809, in _create_c_op
    serialized)
MemoryError: std::bad_alloc
  In call to configurable 'decode' (<function decode at 0x7f81a59d1268>)
  In call to configurable 'decode_from_file' (<function decode_from_file at 0x7f81a59d16a8>)
  In call to configurable 'infer_model' (<function infer_model at 0x7f81a59d4488>)
  In call to configurable 'run' (<function run at 0x7f81a59d4b70>)
```